modules/javafx.graphics/src/main/java/com/sun/javafx/iio/javax/XImageLoader.java

```diff
@@ -0,0 +1,300 @@
+import java.awt.image.ComponentSampleModel;
+import java.awt.image.DataBuffer;
+import java.awt.image.DataBufferByte;
+import java.awt.image.DataBufferInt;
+import java.awt.image.IndexColorModel;
+import java.awt.image.MultiPixelPackedSampleModel;
+import java.awt.image.SinglePixelPackedSampleModel;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.IntBuffer;
+import java.util.Arrays;
+import java.util.Objects;
+
+import static java.awt.image.BufferedImage.*;
+
+public class XImageLoader implements ImageLoader {
+
+    private final ImageReader reader;
+    private final ImageInputStream stream;
+    private final ImageFormatDescription description;
+
+    public XImageLoader(ImageReader reader, ImageInputStream stream) throws IOException {
+        this.reader = reader;
+        this.stream = stream;
+        this.description = new ImageDescriptor(
+            reader.getFormatName(), new String[0], new ImageFormatDescription.Signature[0],
+            reader.getOriginatingProvider() == null ? new String[0] :
+                Arrays.stream(reader.getOriginatingProvider().getMIMETypes())
+                    .map(type -> type.substring(type.indexOf('/')))
+                    .toArray(String[]::new));
+
+        reader.setInput(stream);
+    }
+
+    @Override
+    public ImageFormatDescription getFormatDescription() {
+        return description;
+    }
+
+    @Override
+    public void dispose() {
+        reader.dispose();
+
+        try {
+            if (stream != null) {
+                stream.close();
+            }
+        } catch (IOException ignored) {
+        }
+    }
+
+    @Override
+    public void addListener(ImageLoadListener listener) {
+        var listenerImpl = new LoadListenerImpl(listener);
+        reader.addIIOReadProgressListener(listenerImpl);
+        reader.addIIOReadWarningListener(listenerImpl);
+    }
+
+    @Override
+    public void removeListener(ImageLoadListener listener) {
+        var listenerImpl = new LoadListenerImpl(listener);
+        reader.removeIIOReadProgressListener(listenerImpl);
+        reader.removeIIOReadWarningListener(listenerImpl);
+    }
+
+    @Override
+    public ImageFrame load(int imageIndex, double w, double h, boolean preserveAspectRatio, boolean smooth,
+                           float screenPixelScale, float imagePixelScale) throws IOException {
+        // We currently don't support animated images loaded via ImageIO
+        if (imageIndex != 0) {
+            return null;
+        }
+
+        int width, height;
+        float pixelScale;
+        ImageReadParam param = reader.getDefaultReadParam();
+
+        if (param.canSetSourceRenderSize()) {
+            int imageWidth = reader.getWidth(imageIndex);
+            int imageHeight = reader.getHeight(imageIndex);
+            int[] widthHeight = ImageTools.computeDimensions(
+                (int)(imageWidth * screenPixelScale), (int)(imageHeight * screenPixelScale),
```

Comment:
Again is there a possibility of overflow here?

Reply:
should it be `(int)` (~floor) or `(int)Math.round()` ?